Shape analysis

lucaspena opened the pull request:
OOPSLA (new kt-collapse rule, shape analysis rules, abstracting nil, etc.)

lucaspena committed:
build: uncomment out kore and smt tests, add some missing expected files

lucaspena committed:
add footprint analysis rule. modify matching to handle heap variable

lucaspena committed:
move footprint analysis from unfolding to matching

lucaspena committed:
matching: footprint analysis uses match to find instance of pto on LHS

nishantjr committed:
testlists: dll-vc05: Use kt as part of the search instead of manually specifying it.

lucaspena committed:
matching: #matchStuck is own sort now, results in stuck configuration if fires

lucaspena committed:
matching: remove spurious side condition in matchAssocComm rule causing rule not to fire

lucaspena committed:
begin nullity analysis strategy

lucaspena committed:
add strategy to abstract away LHS spatial part

lucaspena committed:
simplification: simplify substitute-equals-for-equals - call subst directly instead of substPatternsMap

lucaspena committed:
modify nullity analysis, kt-solve-implications

lucaspena committed:
extra dnfPs case: and(not(and))

lucaspena committed:
and-split-rhs added

lucaspena committed:
more general kt-collapse rule subsuming ad-hoc rule

lucaspena committed:
wip: somewhat ad-hoc strategy for abstract nil terms

lucaspena committed:
kt: added fail case for kt-collapse back

lucaspena committed:
simplification: fix small bug in #getNewVariables

lucaspena committed:
kt: ktForEachLRP(.Patterns) fails instead of noop

lucaspena committed:
WIP: knaster-tarski: substituting keeps predicate information after kt-unfold

lucaspena committed:
kt: split kt-unfold into kt-unfold and kt-subst

lucaspena committed:
wip: lift-constraints has incorrect canonical form

lucaspena committed:
completely reimplement lift-constraints (correctly this time)

lucaspena committed:
alpha rename in RHS of kt-subst

nishantjr committed:
incomplete: resolution rule

nishantjr committed:
fix kt-collapse rule

lucaspena committed:
knaster-tarski: dont call dnfPs until kt-collapse-unsat

lucaspena committed:
use substitution for kt-wrap, left-unfold instead of specialized FOL/SL rules

lucaspena committed:
add test for matching over symbolic heap

lucaspena committed:
wip on kt-collapse with resolution

lucaspena committed:
add constructor axiom

lucaspena committed:
check-lhs-constraint-unsat-debug strategy

lucaspena committed:
wip on match-assoc-comm test of match within a conjunction of multiple seps

lucaspena committed:
rename instantiate-axiom to instantiate-heap-axiom

lucaspena committed:
pto-is-injective strategy

lucaspena committed:
wip on abstracting nil

lucaspena committed:
nil no longer parameterized symbol

lucaspena committed:
matching with heap variable must be at end of pattern list

lucaspena committed:
move spatial-patterns-match strategy

lucaspena committed:
footprint-analysis: symbolic heap must be at end of sep

lucaspena committed:
dll-vc07 goes through

xc93 commented:
What's the plan for this PR? It looks too big to be merged given its current status.

nishantjr commented:
Yes, we're putting it here so that the CI runs the tests. We're going to break this apart into multiple PRs later (the subst PR is one of them)

nishantjr committed:
Fix codeblock formatting

lucaspena committed:
right-unfold-all: base case is noop, falls back on search-bound base case

lucaspena committed:
spatial-patterns-match: iterates on all heaps on LHS

lucaspena committed:
search-bound: add spatial-patterns-match before smt call

lucaspena committed:
dll-vc11 goes through with footprint-analysis, abstract-nil

lucaspena committed:
filterVariablesBySort: move to kore-lang

lucaspena committed:
nullity-analysis: generalize to work over all location variables

lucaspena committed:
kt-collapse: failing rule does not rewrite claim cell

lucaspena committed:
kt-collapse-unsat: heap variable added at end of sep

lucaspena committed:
change #hole to VariableName instead of Variable

lucaspena committed:
kt-collapse: use subgoal to collapse context using full power of prover

lucaspena committed:
match: rotate heap variable when not at end of pattern list

lucaspena committed:
match: lhs can have multiple seps

lucaspena committed:
spatial-patterns-equal: wip on multiple seps on LHS

lucaspena committed:
spatial-patterns-match: works with multiple seps anywhere on lhs

lucaspena committed:
spatial-patterns-equal: handles case of multiple seps on RHS

lucaspena committed:
kt-collapse: add heap variable to rule 2a on LHS

lucaspena committed:
matching: match on multiple heaps on RHS

lucaspena committed:
spatial-patterns-equal: lift-constraints after removing heaps

lucaspena committed:
kt-collapse-resolve: still applies with empty constraints

lucaspena committed:
match: terms can be and(.Patterns)

lucaspena committed:
frame: use spatial-patterns-match instead of subsumes-spatial

lucaspena committed:
abstract: more general abstraction of nil, dll-vc17 goes through

lucaspena committed:
re-parameterize nil with sort (needed for nll cases with multiple heaps)

nishantjr committed:
kt-collapse: Remove unneeded sidecondition on Univerals

lucaspena committed:
strategy for nll-vc03

lucaspena committed:
abstract: additional case for #createDisequalities

nishantjr committed:
strategies/kt: Split kt into multiple modular strategies; fix broken kt-resolve strategy

nishantjr committed:
simplification: lift-or: handle \implies(\and(\or()...)

nishantjr committed:
simplification: instantiate-existentials: also instantiate existentials in LHS of context

nishantjr committed:
core: and-split: \and(.Patterns) => success

nishantjr committed:
search-sl: Do a check-lhs-constraint-unsat since the kt-collapse-no-match is likely lhs-unsat

lucaspena committed:
strategy for nll-vc04

nishantjr committed:
Update notes about `kt-collapse`

lucaspena committed:
frame of #hole goes to fail

lucaspena committed:
strategy for nll_vc08 and nll_vc10

nishantjr committed:
lib/render-proof-tree

nishantjr committed:
match-debug

nishantjr committed:
matching: match: Everything matches \top

nishantjr committed:
matching: spatial-patterns-match: Fix non-termination; Use #match

nishantjr committed:
kt: Number of fixes for:

* with-each-implication-context:  Only try collapsing contexts on LHS
* kt-collapse-matched: Rename #hole to a fresh name
* kt-collapse-matched: When proving 2b, remove matching implication contexts from both sides

nishantjr committed:
t/SL-COMP18/bench/qf_shid_entl/tseg_join_tree_entail_tseg.sb.smt2 passing

lucaspena committed:
strategy to remove constraints from lhs, lsegex4_slk-1 goes through

lucaspena committed:
smt-driver: nil is a constructor

lucaspena committed:
simplification: fix comment on remove constraints

nishantjr committed:
Remove tests without expected results

lucaspena committed:
testlists: skl3-vc10 goes through

lucaspena committed:
ltl test: until implies eventually

nishantjr committed:
t/: Place LTL test in own folder

nishantjr committed:
kt: Do not normalize in KT. This happens in the search in any case

nishantjr committed:
begin work on mu/ltl

nishantjr committed:
kore: getReturnSort: Case of \mu, \or, \and

lucaspena committed:
wip on adding support for \mu for ltl example

lucaspena committed:
fix spacing right-unfold-Nth

lucaspena committed:
fix typo

lucaspena committed:
more wip on until implies eventually

nishantjr committed:
build: Generate tests for t/*.smt and t/*.kore (these are not part of the smoke tests however)

nishantjr committed:
build: Replace ls_nonrec_entail_ls_{07 => 05}.sb.smt2 and add kore smoke test

nishantjr committed:
configuration: strategies are now in the <k> cell. goals in <claim>

This fixes issues in the various drivers. This logically makes more sense
since it is the strategy cell that controls the execution.

nishantjr committed:
Fix claim vs strategy cell issues

nishantjr committed:
fix claim v strategy

nishantjr committed:
claim v strategy

nishantjr committed:
claim v strategy

nishantjr committed:
drivers/base: Fix typo in configuration format attribute

nishantjr committed:
build: Work around different extensions being used by kore and smt tests

nishantjr committed:
driver/base: configuration: Claim cell is \bottom rather than top initially to avoid confusion

nishantjr committed:
drivers/smt-driver: Rename `#rest` => `#mutualRecUnfold` to avoid ambiguity

nishantjr committed:
lang/kore: Reimplement #dnf

nishantjr committed:
t/unit/match-assoc: Fix missing { Data }

nishantjr committed:
matching: #match: Fix missing side condition

nishantjr committed:
!!! t/unit/match-assoc-comm: Regression test for match failing when it shouldn't be

lucaspena committed:
matching over mu

nishantjr committed:
kore-helpers :#liftOr: combine multiple bases cases by using isDnfAtom

nishantjr committed:
t/unit/match-assoc-comm: Test should be calling #match, not #matchAssocComm

lucaspena committed:
fix bugs in dnf, wip on until implies eventually

lucaspena committed:
fix preprocessing for matching, until implies eventually goes through

lucaspena committed:
wip on gfp strategy for nu, base case for ltl ind rule goes through

nishantjr committed:
kore: Extend subst and getFreeVariables to \nu

nishantjr committed:
kore: #nnf handles \exists { .Patterns }

nishantjr committed:
strategies/core: rhs-top: Add failure case

nishantjr committed:
strategy: constradiction: Look for a contradiction between the first clause and any other clause

nishantjr committed:
matching: #matchAssoc: Syntactically match on `\not`

nishantjr committed:
matching: Use #nnf to allow handeling more cases such as \implies

nishantjr committed:
matching: #matchAssoc: Patterns may use set variables too

nishantjr committed:
t/ltl/always-propagates.kore

lucaspena committed:
fix bugs in dnf, remove need for liftconstraints

nishantjr committed:
Update README

lucaspena committed:
nu is not a predicate pattern

lucaspena committed:
fix match case with error, ind goes through again

lucaspena committed:
render-proof-tree env specifies python3

lucaspena committed:
render-proof-tree: goals no longer have active cell, first goal is active

lucaspena committed:
kt-collapse: fix bug where substitution in match case is not applied fully

nishantjr committed:
build: Uniform extensions for all tests
